src/common/dns_utils.cpp

```diff
@@ -323,4 +324,106 @@ bool DNSResolver::check_address_syntax(const char *addr) const
+        return s.substr(pos, 95);
+      }
+    }
+    return std::string();
+  }
+  /**
+   * @brief gets a monero address from the TXT record of a DNS entry
+   *
+   * gets the monero address from the TXT record of the DNS entry associated
+   * with <url>.  If this lookup fails, or the TXT record does not contain an
+   * XMR address in the correct format, returns an empty string.  <dnssec_valid>
+   * will be set true or false according to whether or not the DNS query passes
+   * DNSSEC validation.
+   *
+   * @param url the url to look up
+   * @param dnssec_valid return-by-reference for DNSSEC status of query
+   *
+   * @return a monero address (as a string) or an empty string
+   */
+  std::vector<std::string> addresses_from_url(const std::string& url, bool& dnssec_valid)
+  {
+    std::vector<std::string> addresses;
+    // get txt records
+    bool dnssec_available, dnssec_isvalid;
+    std::string oa_addr = DNSResolver::instance().get_dns_format_from_oa_address(url);
+    auto records = DNSResolver::instance().get_txt_record(oa_addr, dnssec_available, dnssec_isvalid);
+
+    // TODO: update this to allow for conveying that dnssec was not available
+    if (dnssec_available && dnssec_isvalid)
+    {
+      dnssec_valid = true;
+    }
+    else dnssec_valid = false;
+
+    // for each txt record, try to find a monero address in it.
+    for (auto& rec : records)
+    {
+      std::string addr = address_from_txt_record(rec);
+      if (addr.size())
+      {
+        addresses.push_back(addr);
+      }
+    }
+    return addresses;
+  }
+}
+
+std::string get_account_address_as_str_from_url(const std::string& url, bool& dnssec_valid)
+{
+  // attempt to get address from dns query
+  auto addresses = detail::addresses_from_url(url, dnssec_valid);
+  if (addresses.empty())
+  {
+    LOG_ERROR("wrong address: " << url);
+    return {};
+  }
+  // for now, move on only if one address found
+  if (addresses.size() > 1)
+  {
+    LOG_ERROR("not yet supported: Multiple Monero addresses found for given URL: " << url);
+    return {};
+  }
+  // inform user of DNSSEC validation status.
+  LOG_PRINT_L0("For URL: " << url << ", DNSSEC validation " << (dnssec_valid ? "passed" : "was unsuccessful, this address may not be correct!"));
+  LOG_PRINT_L0("Monero Address = " << addresses[0]);
+  return addresses[0];
+}
+
+bool get_account_address_from_str_or_url(
+    cryptonote::account_public_address& address
+  , bool& has_payment_id
+  , crypto::hash8& payment_id
+  , bool testnet
+  , const std::string& str_or_url
+  )
+{
+  if (cryptonote::get_account_integrated_address_from_str(address, has_payment_id, payment_id, testnet, str_or_url))
+    return true;
+  bool dnssec_valid;
+  std::string address_str = get_account_address_as_str_from_url(str_or_url, dnssec_valid);
+  return !address_str.empty() &&
+          cryptonote::get_account_integrated_address_from_str(address, has_payment_id, payment_id, testnet, address_str);
+}
+
 }  // namespace tools
```

Comment:
I appreciate if you could take care of organizing the code. Your PR #1468 seems to have a lot of overlap with this PR, anyway.